features/dd-sdk-android-trace/src/test/kotlin/com/datadog/opentracing/SpanCleanerTest.kt

```diff
@@ -0,0 +1,184 @@
+        // Given
+        val pendingTraces = forge.aList(size = forge.anInt(min = 10, max = 50)) { mock<PendingTrace>() }
+        pendingTraces.forEach {
+            testeCleaner.addPendingTrace(it)
+        }
+        assertThat(testeCleaner.pendingTraces.values).containsExactlyInAnyOrderElementsOf(pendingTraces)
+
+        // When
+        testeCleaner.close()
+
+        // Then
+        pendingTraces.forEach {
+            verify(it).clean()
+        }
+    }
+
+    // endregion
+
+    // region `equals` contract
+
+    @Test
+    fun `M contain only one entry W add same PendingTrace twice`(forge: Forge) {
+        // Given
+        val pendingTrace = mockPendingTrace(forge)
+        testeCleaner.addPendingTrace(pendingTrace)
+        testeCleaner.addPendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testeCleaner.pendingTraces).hasSize(1)
+        assertThat(testeCleaner.pendingTraces.values).containsExactly(pendingTrace)
+    }
+
+    @Test
+    fun `M contain only one entry W add {same PendingTrace mutated, twice}`(forge: Forge) {
+        // Given
+        val pendingTrace = mockPendingTrace(forge)
+        testeCleaner.addPendingTrace(pendingTrace)
+        pendingTrace.addSpan(mock())
+        pendingTrace.addSpan(mock())
+        testeCleaner.addPendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testeCleaner.pendingTraces).hasSize(1)
+        assertThat(testeCleaner.pendingTraces.values).containsExactly(pendingTrace)
+
+        // When
+        testeCleaner.removePendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testeCleaner.pendingTraces).isEmpty()
+    }
+
+    // endregion
+
+    // region concurrency
+
+    @Test
+    fun `M not leak any PendingTrace W add, mutate and remove { different threads }`(forge: Forge) {
+        // Given
+        val pendingTraces = forge.aList(size = forge.anInt(min = 10, max = 50)) { mockPendingTrace(forge) }
+        val addCountDownLatch = CountDownLatch(pendingTraces.size)
+        val removeCountDownLatch = CountDownLatch(pendingTraces.size)
+        pendingTraces.take(pendingTraces.size).forEach { pendingTrace ->
+            Thread {
+                testeCleaner.addPendingTrace(pendingTrace)
+                repeat(forge.anInt(min = 1, max = 10)) {
+                    pendingTrace.addSpan(mock())
+                }
+                addCountDownLatch.countDown()
+            }.start()
+        }
+        pendingTraces.takeLast(pendingTraces.size / 2).forEach { pendingTrace ->
+            Thread {
+                if (forge.aBool()) {
+                    repeat(forge.anInt(min = 1, max = 10)) {
+                        pendingTrace.addSpan(mock())
+                    }
+                }
+                testeCleaner.addPendingTrace(pendingTrace)
+                repeat(forge.anInt(min = 1, max = 10)) {
+                    pendingTrace.addSpan(mock())
+                }
+                addCountDownLatch.countDown()
```

Comment:
there is a small difference with a block above, it is using `if (forge.aBool())`, is it intended?

Reply:
yeah I just wanted to test random case where the pending trace has something before was added, maybe should I add 2 tests here ? Isn't going to be too much ?